core/discord_auth.py

```diff
@@ -0,0 +1,115 @@
+import functools
+import typing
+
+from django.http import HttpRequest, HttpResponse
+from django.shortcuts import redirect
+from django.urls import reverse
+from core.discord_api import (
+    DISCORD_REFRESH_COOKIE,
+    DISCORD_TOKEN_COOKIE,
+    fetch_user,
+    DISCORD_ID_COOKIE,
+)
+from core.models import User, UserRole
+
+
+def user_logged_in(request: HttpRequest) -> User | None:
+    """Returns user if logged in. If not, returns 0 when session should be refreshed or 1 when session is invalid."""
+    token = request.COOKIES.get(DISCORD_TOKEN_COOKIE)
+    if token is None:
+        return None
+
+    discord_id = request.COOKIES.get(DISCORD_ID_COOKIE)
+    if discord_id is None:
+        user = fetch_user(token)
```

Comment:
Maybe extract `fetch_user` above and just compare it with `User.objects` when discord_id is not none